### ISSUE TITLE
[RTC] Better mechanism for generating and inlining clientlibs

### ISSUE THREAD

**Issue by vladbailescu** [adobe/aem-core-wcm-components#950]
# Problem

The recent contribution for AMP has exposed a need for a better way to generate clientlibs and, possibly, inline them (AMP and Email requires CSS to be inlined).

Adobe already has support for handling clientlibs (especially compiling/concatenating them) so we should try to reuse as much as we can.

The normal way to include clientlibs now in a page is to call the Granite-provided helper:

```html
<sly data-sly-use.clientlib="/libs/granite/sightly/templates/clientlib.html"
     data-sly-call="${clientlib.all @ categories = [..]}">
```

However, this is not flexible enough to allow inlining or custom attributes (`async`, `crossorigin`, `media` etc)

# Proposed Solution

We should provide a better clientlib handler that allows inlining and custom attributes. It would replace the existing `/libs/granite/sightly/templates/clientlib.html` and it's associated `ClientLibUseObject `. It will be using https://helpx.adobe.com/experience-manager/6-5/sites/developing/using/reference-materials/javadoc/com/adobe/granite/ui/clientlibs/HtmlLibraryManager.html to ensure actual handling of clientlibs contents happens the same way as it does in AEM and could be at some point integrated into AEM.

**For AMP and Email**, we should create a provider for the clientlib categories that are actually used on the page, so that we keep the inlined/referenced CSS/JS slim.

# Usage Examples

```html
<sly data-sly-use.clientlib=".../clientlib.html" />
<!--/* Inlining CSS, JS or both: */-->
<sly data-sly-call="${clientlib.js  @ categories=['cat1', 'cat2'], inline=true}"/>
<sly data-sly-call="${clientlib.css @ categories=['cat1', 'cat2'], inline=true}"/>
<sly data-sly-call="${clientlib.all @ categories=['cat1', 'cat2'], inline=true}"/>
<!--/* Passing attributes to the CSS or JS elements: */-->
<sly data-sly-call="${clientlib.js  @ categories=['cat1', 'cat2'],
        attributes={async:true, crossorigin:'anonymous'}}"/>
<sly data-sly-call="${clientlib.css @ categories=['cat1', 'cat2'],
        attributes={media:'print'}}"/>
<sly data-sly-call="${clientlib.all @ categories=['cat1', 'cat2'],
        cssAttributes={media:'print'}, jsAttributes={async:true}}"/>
```

CC @gabrielwalt 

# Similar Solutions
* @nateyolles -  https://github.com/nateyolles/aem-clientlib-async

**Comment by gabrielwalt** [comment]
Note that in Vlad's usage examples, the way the object literal is written would require the following HTL feature: https://github.com/adobe/htl-spec/issues/63
Currently, this object needs to be created programmatically by a Java or JS use object.

For convenience, in addition to the `attributes` property, it might make sense to also support the 4 attributes that Nate's solution has also as direct template parameters:
```html
<sly data-sly-call="${clientlib.js  @ categories=['cat1', 'cat2'], async=true}"/>
<sly data-sly-call="${clientlib.js  @ categories=['cat1', 'cat2'], defer=true}"/>
<sly data-sly-call="${clientlib.js  @ categories=['cat1', 'cat2'], onload='myFunction()'}"/>
<sly data-sly-call="${clientlib.all @ categories=['cat1', 'cat2'], crossorigin='anonymous'}"/>
```

**Comment by stefanseifert** [comment]
another comparable solution with a bit more features than that from nateyolles is 
https://wcm.io/wcm/ui/clientlibs
https://wcm.io/wcm/ui/clientlibs/usage.html
https://github.com/wcm-io/wcm-io-wcm/tree/develop/ui/clientlibs

it also wraps the HtmlLibraryManager.

**Comment by justinedelson** [comment]
FWIW, inlining can be done using the existing rewriter pipeline in AEM. See https://github.com/Adobe-Consulting-Services/acs-aem-commons/blob/master/bundle/src/main/java/com/adobe/acs/commons/rewriter/impl/StylesheetInlinerTransformerFactory.java

IIRC, this was created for email use cases.

**Comment by gabrielwalt** [comment]
FYI, in order to reduce the duplication of existing open-source efforts, we're intending to contribute the enhancement described in this RTC to https://wcm.io/wcm/ui/clientlibs, and to use that wcm.io clientlibs extension for the Core Components.

**Comment by brunob15** [comment]
Is there currently a way to avoid duplicate clientlib imports when importing the clientlibs from the components' template?

**Comment by vladbailescu** [comment]
Solved with #1084
